notebook/exploration-reports/2021.09-dag-pb-encryption.md

```diff
@@ -0,0 +1,202 @@
+
+Note: An IV of 0 is acceptable in this scenario because we use unique keys for every string of data. Another approach would be to use extra bits from the derivation hash (or extra random bits) but that is unnecessary.
+
+## DAG Encryption
+
+An alternative to the DAG-PB format will be created that enables encryption. Note that we can not just add fields to the existing message because [the spec forbids additional fields](https://github.com/ipld/specs/blob/master/block-layer/codecs/dag-pb.md#protobuf-strictness).  Otherwise this is mostly an addition to the existing fields. If that messaging is updated we could preserve the old Name and Data fields this would allow backwards compatibility for pinning.
+
+DAG-PB-ENC is the new DAG format that supports optional encryption.
+
+```proto
+message Link {
+  // binary CID (with no multibase prefix) of the target object
+  optional bytes Hash = 1;
+
+  // The Name field has been moved to the EncryptedNode.
+  reserved "Name";
+  reserved 2;
+
+  // The TSize field has been moved to the EncryptedNode as CumulativeSize.
+  reserved "TSize";
+  reserved 3;
+}
+
+// Information that is not required for pinning which may be encrypted.
+message EncryptedLink {
+  optional string Name = 2;
+  optional bytes key = 4;
+}
+
+message EncryptedNode {
+  repeated EncryptedLink = 2;
+
+  // Opaque user data
+  optional bytes Data = 1;
+
+  // The cumulative size of the target and all of its children.
+  optional uint64 CumulativeSize = 3;
+}
+
+message Node {
+  // refs to other objects
+  repeated Link Links = 2;
+
+  enum EncryptionMode {
+    None = 0;
+    AES256CBC = 1;
+  }
+  optional EncryptionMode Encryption = 3;
+
+  // A serialized EncryptedNode proto encrypted using the method specified by EncryptionMode.
+  optional bytes EncryptedNode = 4;
+
+  // The Data field has been moved to the EncryptedNode.
+  reserved "Data";
+  reserved 1;
+}
+```
+
+This proto will be used the same way as the [DAG-PB](https://github.com/ipld/specs/blob/master/block-layer/codecs/dag-pb.md) proto with the following differences.
+
+- When Encryption != None and the encryption key is not available, only the Hash of the links are available. Implementations should handle this gracefully and can not allow traversal by name.
+- While not required by the protocol itself implementations MUST sort the `links` elements by the CID. This ensures that information about the order of links is not leaked.
+
+## References
+
+### For Pinning
+
+If you want to pin your content without providing access to the data to the pinner you can pin the base CID as usual.
+
+### For Sharing
+
+A text format reference is formed by appending information about the encryption to the CID. This takes the form of a hyphen "-", followed by the multibase encoding of the key. The algorithm is not specified in the key but in the DAG node.
+
+#### Example
+
+QmVhiZNnvhqrbTLRsh6JnryJ5eTUwygTjTg5hUBKAfeP1H-z7Z9ZajGKvb6C6LaiB7fnsWQZNwq8roEKCFdt3Fbb1qt9
```

Comment:
This is a tricky one, there's so much infrastructure baked into the CID pattern, where it either needs to be a valid CID, or a CID + a valid path (`/like/this`), otherwise these strings don't propagate through the various layers. So adding an arbitrary extension to a CID is going to be a difficult thing to make happen.
There's always handwaving we do about CIDv2 that may allow valid extensions for various reasons, but it really is just handwaving and thought-bubbling for now.

Reply:
Yeah, this will definitely break some assumptions. If you have a solution that could make this easier I'm open to ideas. I'm not particular on the exact format. But the "magic" in this design is that you need to pass a key in to get the decrypted data out.